### PR TITLE
[LLVM][SelectionDAG] Remove scalable vector restriction from poison analysis.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
[LLVM][SelectionDAG] Remove scalable vector restriction from poison analysis.
    
The following functions have an early exit for scalable vectors:
  SelectionDAG::canCreateUndefOrPoison
  SelectionDAG:isGuaranteedNotToBeUndefOrPoison
    
The implementations of these don't look to be sensitive to the
vector type other than some uses of demanded elts analysis that
doesn't fully support scalable types.  That said the initial
calculation demands all elements and so I've followed the same
scheme as used by SelectionDAG::isSplatValue.
    
If this is overly aggressive I can instead move the scalable
vector based exit deeper into these functions at the point where
the demanded elts analysis is used.

### CODE REVIEW

**Comment by paulwalker-arm** [comment]
I don't fully understand the RISCV changes but when investigating vwadd_vx_splat_sext_i1 it looks like a target specific DAG combine is now firing for `SUB x, SPLAT(imm)` which was previously blocked because a freeze sat between the SUB and SPLAT.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 1]
```diff
@@ -5140,14 +5140,11 @@ bool SelectionDAG::isGuaranteedNotToBeUndefOrPoison(SDValue Op, bool PoisonOnly,
...
-  // TODO: Assume we don't know anything for now.
   EVT VT = Op.getValueType();
-  if (VT.isScalableVector())
-    return false;
-
-  APInt DemandedElts = VT.isVector()
-                           ? APInt::getAllOnes(VT.getVectorNumElements())
-                           : APInt(1, 1);
+  APInt DemandedElts =
+      VT.isVector() ? APInt::getAllOnes(
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5145]: SimplifyDemandedBits does the same thing but writes it as

```
  APInt DemandedElts = VT.isFixedLengthVector()                                  
                           ? APInt::getAllOnes(VT.getVectorNumElements())        
                           : APInt(1, 1);
```

Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5145]: That's nicer, thanks, I'll copy that version instead.

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll [hunk 2]
```diff
@@ -1501,18 +1501,23 @@ define <vscale x 8 x i32> @vwadd_vx_splat_zext_i1(<vscale x 8 x i1> %va, i16 %b)
 ; RV32:       # %bb.0:
 ; RV32-NEXT:    slli a0, a0, 16
 ; RV32-NEXT:    srli a0, a0, 16
-; RV32-NEXT:    vsetvli a1, zero, e32, m4, ta, mu
+; RV32-NEXT:    vsetvli a1, zero, e32, m4, ta, ma
 ; RV32-NEXT:    vmv.v.x v8, a0
-; RV32-NEXT:    vadd.vi v8, v8, 1, v0.t
+; RV32-NEXT:    addi a0, a0, 1
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll:1506]: (not blocking discussion for this review)

@topperc What's your feeling on this change?  This looks like either the before or after are about equal to me, do you agree?

File: llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll [hunk 3]
```diff
@@ -1501,18 +1501,23 @@ define <vscale x 8 x i32> @vwadd_vx_splat_zext_i1(<vscale x 8 x i1> %va, i16 %b)
...
 ; RV64-NEXT:    srli a0, a0, 48
-; RV64-NEXT:    vsetvli a1, zero, e32, m4, ta, mu
+; RV64-NEXT:    vsetvli a1, zero, e16, m2, ta, ma
+; RV64-NEXT:    vmv.v.x v12, a0
+; RV64-NEXT:    vsetvli zero, zero, e32, m4, ta, ma
 ; RV64-NEXT:    vmv.v.x v8, a0
-; RV64-NEXT:    vadd.vi v8, v8, 1, v0.t
+; RV64-NEXT:    li a0, 1
+; RV64-NEXT:    vsetvli zero, zero, e16, m2, ta, mu
+; RV64-NEXT:    vwaddu.vx v8, v12, a0, v0.t
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll:1520]: (non blocking for this review, continuing from previous)

This one looks a bit more questionable.  It looks like maybe we need a guard in the vwadd combine for the case where the RHS is a legal immediate?   It'd be really useful here if we have a vwadd.vi form, but we don't.  Ignoring the passthru issue, which form do you think is likely better - vwaddu.vx w/immediate in register or vzext.vf + vadd.vi?

File: llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll [hunk 4]
```diff
@@ -1570,20 +1575,23 @@ define <vscale x 8 x i32> @vwadd_vx_splat_sext_i1(<vscale x 8 x i1> %va, i16 %b)
...
 ; RV64-NEXT:    srai a0, a0, 48
-; RV64-NEXT:    vsetvli a1, zero, e32, m4, ta, mu
+; RV64-NEXT:    vsetvli a1, zero, e16, m2, ta, ma
+; RV64-NEXT:    vmv.v.x v12, a0
+; RV64-NEXT:    vsetvli zero, zero, e32, m4, ta, ma
 ; RV64-NEXT:    vmv.v.x v8, a0
 ; RV64-NEXT:    li a0, 1
-; RV64-NEXT:    vsub.vx v8, v8, a0, v0.t
+; RV64-NEXT:    vsetvli zero, zero, e16, m2, ta, mu
+; RV64-NEXT:    vwsub.vx v8, v12, a0, v0.t
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll:1594]: (non blocking for this review, continuing from previous)

In this case, the rv32 codegen looks clearly better than the rv64.

**Review by preames — APPROVED** [review]
LGTM.  I've audited the code, and don't spot any obvious problems, and the general approach is good.  

You probably want to add handling for SPLAT_VECTOR in getValidShiftAmountRange

File: llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll (hunk 2 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll:1506]: Agreed it looks about equal

File: llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll (hunk 3 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/vwadd-sdnode.ll:1520]: vzext.vf + vadd.vi seems like more work in the vector ALU. So I think vwaddu.vx w/immediate is better.

Hopefully we can find a way to fix this to match the RV32 codegen.

**Comment by preames** [comment]
> > You probably want to add handling for SPLAT_VECTOR in getValidShiftAmountRange
> 
> I've taken a look and think getValidShiftAmountRange already works? It'll drop into `computeKnownBits` which already handles SPLAT_VECTOR.

Looks like you're right.  I glanced at the BUILD_VECTOR specific code in getValidShiftAmountRange and without looking at it real closely assumed we'd need a parallel.  Looking at it now, I see that's computing a constant range directly from constants instead of getting a less precise one from KnownBits.  SPLAT_VECTOR doesn't have that challenge since there's only a single value, so if the operand is constant the knownbits will be precise.

**Pull request merged**